Resolved: Missing Sound Effect on Gold Drop Warning #319

AstralSkies opened the pull request:
- Added "Login" sound effect to gold drop warning dialog.
- Ensured single occurrence of warning message and sound, aligning with Vanilla EO.

ethanmoffat commented on EndlessClient/Dialogs/Factories/ItemTransferDialogFactory.cs:
```diff
@@ -44,7 +46,19 @@ public ItemTransferDialog CreateItemTransferDialog(string itemName, ItemTransfer
                 message);
-            dlg.DialogClosing += (_, _) => _sfxPlayer.PlaySfx(SoundEffectID.DialogButtonClick);
+
+            dlg.DialogClosing += (_, _) =>
```
`ItemTransferDialogFactory` shouldn't have/need any changes, the sound effect should be added to the controller where the `warningMsg` dialog is created and shown.

This class should stay generic to creating a dialog for transferring items somewhere, not have special logic for gold drop.

AstralSkies replied:
Sure, I'll do that 👍

ethanmoffat commented on EndlessClient/Dialogs/Factories/ItemTransferDialogFactory.cs:
```diff
@@ -44,7 +46,19 @@ public ItemTransferDialog CreateItemTransferDialog(string itemName, ItemTransfer
+
+            dlg.DialogClosing += (_, _) =>
+            {
+                if (itemName == "Gold" && !_goldSoundPlayed)
```
When you move this logic to the controller, you can add the sound effect to when the "safety comment" message dialog is opened rather than when the item transfer is closed. I think that makes a bit more sense.

AstralSkies commented:
Hi Ethan,

Quick update: the recent unusual commit history was due to my sister experimenting with Git on my project. I’ve cleaned up the history and addressed all requested changes in the latest commit (96cb547). The PR is now ready for your review again.

ethanmoffat commented on EndlessClient/Dialogs/Factories/ItemTransferDialogFactory.cs:
```diff
@@ -44,13 +44,21 @@ public ItemTransferDialog CreateItemTransferDialog(string itemName, ItemTransfer
                 message);
-            dlg.DialogClosing += (_, _) => _sfxPlayer.PlaySfx(SoundEffectID.DialogButtonClick);
+
+            if (playDefaultSound)
```
It looks like the official client doesn't play any sound for the "ok" button. I checked "drop", "junk", "buy", and "sell". I'm good with either:

1) Only play the `DialogButtonClick` sound on "cancel" button press
2) Play `DialogButtonClick` on either button. Play `DialogButtonClick` and safety comment message together

I don't really care too much about the sounds overlapping in this case. I'd rather avoid having a branching arg in the method signature for which sound is played.

That means the options for this file are:
1) Undo all the changes, then check which button was pressed and only play the default sfx if the button is "cancel"
2) Undo all the changes and leave this file untouched

AstralSkies replied:
I'd rather go for 1) as that's what I initially made me include playDefaultSound: 

It would mean the inclusion of a new using statement along with 
![image](https://github.com/ethanmoffat/EndlessClient/assets/70440468/d6bebe92-3e19-4028-bb2b-d1d4c4b1c35b)

Would this be okay?

Best,
Stephen

ethanmoffat commented on EndlessClient/Controllers/InventoryController.cs:
```diff
@@ -50,6 +51,9 @@ public class InventoryController : IInventoryController
         private readonly ILocalizedStringFinder _localizedStringFinder;
+        private readonly ISfxPlayer _sfxPlayer;
+
+        private bool _goldWarningShown = false;
```
Now that this class contains state, you will need to change `[AutoMappedType]` to `[AutoMappedType(IsSingleton = true)]`

AstralSkies replied:
![image](https://github.com/ethanmoffat/EndlessClient/assets/70440468/62faa6a2-7376-4d1c-9843-7ed331274e8d)

ethanmoffat commented on EndlessClient/Controllers/InventoryController.cs:
```diff
@@ -348,29 +354,41 @@ public void TradeItem(EIFRecord itemData, InventoryItem inventoryItem)
+
             if (inventoryItem.Amount > 1)
             {
+                if (inventoryItem.ItemID == 1 && inventoryItem.Amount > 10000 && transferType == ItemTransferDialog.TransferType.DropItems)
```
If you want to break the condition out of the if statement, do something like

`var isLargeGoldItemDrop = inventoryItem.ItemID == 1 && inventoryItem.Amount > 10000 && transferType == ItemTransferDialog.TransferType.DropItems;`

I think this should also move closer to where it's used since I'm asking you to remove the `playDefaultSound` parameter from the item transfer dialog factory too.

AstralSkies replied:
![image](https://github.com/ethanmoffat/EndlessClient/assets/70440468/7a0f94fb-0cf2-4065-81dc-811c8082b70a)

ethanmoffat commented:
LGTM, thanks for sticking with it through the feedback cycle. Much appreciated!
